File: rust-rsc-prototype/app/catalog/js/product/[id]/page.js
const dataOrigin = process.env.CATALOG_DATA_ORIGIN || 'http://127.0.0.1:3041'
const benchmarkMode = process.env.CATALOG_BENCHMARK_MODE === '1'
const data = require('../../../../../catalog/query-data')
import CatalogControls from '../../../controls'

export default async function JavaScriptProduct({ params }) {
  const { id } = await params
  const product = benchmarkMode
    ? await fetch(`${dataOrigin}/product/${encodeURIComponent(id)}`, {
        cache: 'no-store',
        signal: AbortSignal.timeout(5000),
      }).then((response) => (response.ok ? response.json() : null))
    : data.product(id)
  if (!product)
    return (
      <main data-product-missing={id}>
        <h1>Product not found</h1>
      </main>
    )
  return (
    <CatalogControls basePath="/catalog/js">
      <ProductDetail implementation="js" product={product} />
    </CatalogControls>
  )
}

function ProductDetail({ implementation, product }) {
  return (
    <article
      className="catalog-detail"
      data-product-detail={product.id}
      data-catalog={implementation}
    >
      <a href="/catalog/js" data-catalog-navigation>
        ← Back to catalog
      </a>
      <p className="catalog-eyebrow">{product.category}</p>
      <h1>{product.name}</h1>
      <p className="catalog-part">Part {product.id}</p>
      <dl>
        <div>
          <dt>Material</dt>
          <dd>{product.material}</dd>
        </div>
        <div>
          <dt>Finish</dt>
          <dd>{product.finish}</dd>
        </div>
        <div>
          <dt>Size</dt>
          <dd>{product.size}</dd>
        </div>
        <div>
          <dt>Available</dt>
          <dd>{product.available}</dd>
        </div>
      </dl>
      <p className="catalog-price">${(product.priceCents / 100).toFixed(2)}</p>
      <form>
        <label>
          Quantity{' '}
          <input name="quantity" type="number" min="1" defaultValue="1" />
        </label>
        <button type="button">Add to order</button>
      </form>
    </article>
  )
}

File: rust-rsc-prototype/app/catalog/rust/product/[id]/page.rs
use next_rsc::{Node, PageProps, RenderError, RenderResult, element};

pub fn render(props: PageProps) -> RenderResult {
    let id = props.params.require("id")?;
    let number = id
        .strip_prefix("RC-")
        .and_then(|value| value.parse::<usize>().ok())
        .filter(|value| (1..=720).contains(value))
        .ok_or_else(RenderError::not_found)?;
    let index = number - 1;
    let categories = [
        ("fasteners", "Fasteners"),
        ("electrical", "Electrical"),
        ("plumbing", "Plumbing"),
        ("material-handling", "Material Handling"),
        ("safety", "Safety"),
        ("machining", "Machining"),
    ];
    let materials = ["Zinc Steel", "Stainless Steel", "Aluminum", "Brass"];
    let finishes = ["Plain", "Black Oxide", "Galvanized", "Anodized"];
    let (category, category_name) = categories[index % categories.len()];
    let name = format!("{category_name} {:03}", (index % 120) + 1);
    let field = |name: &str, value: String| {
        element(
            "div",
            [
                element("dt", [Node::text(name)]),
                element("dd", [Node::text(value)]),
            ],
        )
    };
    Ok(element(
        "article",
        [
            element("a", [Node::text("← Back to catalog")])
                .prop("href", "/catalog/rust")
                .prop("data-catalog-navigation", true),
            element("p", [Node::text(category)]).prop("className", "catalog-eyebrow"),
            element("h1", [Node::text(name)]),
            element("p", [Node::text(format!("Part {id}"))]).prop("className", "catalog-part"),
            element(
                "dl",
                [
                    field("Material", materials[index % 4].to_owned()),
                    field("Finish", finishes[(index / 3) % 4].to_owned()),
                    field("Size", format!("{} mm", (index % 24) + 1)),
                    field("Available", (8 + ((index * 29) % 940)).to_string()),
                ],
            ),
            element(
                "p",
                [Node::text(format!(
                    "${:.2}",
                    (175 + ((index * 137) % 18500)) as f64 / 100.0
                ))],
            )
            .prop("className", "catalog-price"),
            element(
                "form",
                [
                    element(
                        "label",
                        [
                            Node::text("Quantity "),
                            element("input", [])
                                .prop("name", "quantity")
                                .prop("type", "number")
                                .prop("min", "1")
                                .prop("value", "1"),
                        ],
                    ),
                    element("button", [Node::text("Add to order")]).prop("type", "button"),
                ],
            ),
        ],
    )
    .prop("className", "catalog-detail")
    .prop("data-product-detail", id)
    .prop("data-catalog", "rust"))
}
